### PR TITLE
RHCLOUD-47250 - MCP S5 - Add tool to check user access without org admin

### PR DESCRIPTION
**Pull request by EvanCasey13**
## Link(s) to Jira
https://redhat.atlassian.net/browse/RHCLOUD-47250

## Description of Intent of Change(s)
  Adds a new MCP tool guide_user_access_delegation. The tool provides comprehensive
  guidance including user verification, role discovery, permission analysis, and actionable next steps for both V1 and
  V2 organizations.

  Changes

  rbac/management/mcp_views.py
  - New guide_user_access_delegation tool registered with @register_tool decorator
  - Accepts username parameter to analyze a specific user
  - Returns comprehensive response with:
    - user_info - User existence and org admin status (via PrincipalProxy)
    - role_info - Details about the "User Access administrator" system role
    - role_access - Permissions granted by the role (rbac::)
    - groups_with_role (V1) - Groups that have this role assigned
    - role_bindings_with_role (V2) - Role bindings granting this role
    - user_already_has_role - Whether the target user already has access
    - user_groups_with_role (V1) / user_role_bindings (V2) - How user has the role
    - org_version - "v1" or "v2" organization type
    - recommendation - Actionable guidance with CAN/CANNOT do lists and next steps
  - Uses direct ORM queries for performance (avoids API call overhead)
  - Detects V2 organizations using is_v2_write_activated(tenant)
  - V1: Queries Group/Policy for role assignments
  - V2: Queries RoleBinding/RoleBindingPrincipal/RoleBindingGroup

  tests/management/test_mcp_views.py
  - New MCPGuideUserAccessDelegationTests class (V1) with 20 test cases:
    - Success and basic structure validation
    - Role discovery and permission display
    - Groups with role detection
    - User already has role detection
    - Recommendation content (CAN/CANNOT do, next steps)
    - Error handling (missing user, missing role, no auth)
    - Edge cases (no groups with role, multiple groups, case-insensitive username)
    - Tool registration in tools/list
  - New MCPGuideUserAccessDelegationV2Tests class with 12 test cases:
    - V2 org version detection
    - Role binding discovery
    - Direct and group-based role binding detection
    - V2-specific next steps (create_role_bindings guidance)
    - Edge cases (no bindings, multiple bindings, binding counts)

  Example Output

  {
    "scenario": "Delegate user access management without Org Admin",
    "username": "dana",
    "org_version": "v1",
    "user_info": {"username": "dana", "is_org_admin": false, ...},
    "role_info": {"name": "User Access administrator", "uuid": "...", "system": true},
    "role_access": {"permissions": ["rbac:*:*"], "permission_summary": "rbac:*:* — create/read/update/delete on all RBAC
   resources"},
    "groups_with_role": [{"name": "Access Governance", "principal_count": 2, ...}],
    "user_already_has_role": false,
    "recommendation": "[V1 Organization] The role you want is 'User Access administrator'..."
  }

## Local Testing
How can the feature be exercised?

  # Run all guide_user_access_delegation tests (V1 + V2)

```
  pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPGuideUserAccessDelegationTests
  tests.management.test_mcp_views.MCPGuideUserAccessDelegationV2Tests
```

  # Run only V1 tests
```
  pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPGuideUserAccessDelegationTests
```

  # Run only V2 tests
```
  pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPGuideUserAccessDelegationV2Tests
```

  # Run a single test
```
  pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPGuideUserAccessDelegationTests.test_guide_user_acce
  ss_delegation_user_already_has_role
```

How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/mcp_views.py [hunk 1]
```diff
@@ -3424,6 +3425,303 @@ def create_cross_account_request(
     return _call_view_write(request, _cross_account_create_view, path, body)
 
 
+@register_tool(
+    description=(
+        "Guide for delegating user access management without making someone an Org Admin. "
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: We need also need it for v2 org - like with v2 roles.

File: rbac/management/mcp_views.py [hunk 2]
```diff
@@ -3424,6 +3425,303 @@ def create_cross_account_request(
...
+        "Provide a username (e.g., 'dana') to get a complete analysis including: "
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: I believe lot of those steps can be done by other tools, can you reduce this tool and add more instructions into description for tool - how to use it -  yes this is pretty concrete use case but let's leverage what is here.

**Comment by lpichler** [comment]
## Code Review Findings

### 1. [Medium] MCPUpdateToolsV2Tests setUp reordering breaks bootstrap (Info)

**File**: `tests/management/test_mcp_views.py` (diff hunk at original line ~4795)

The refactoring of `MCPUpdateToolsV2Tests.setUp` reorders `TenantMapping.objects.create()` to run BEFORE `V2TenantBootstrapService(NoopReplicator()).bootstrap_tenant(self.tenant)`. Because `bootstrap_tenant()` checks for an existing `TenantMapping` via `try_lock_tenant_for_bootstrap()` and returns early if one exists, the bootstrap becomes a **no-op**. This means:

- No ROOT workspace is created
- No DEFAULT workspace is created
- No seeded V2 roles are created
- No built-in relationships are established

Tests within `MCPUpdateToolsV2Tests` that depend on `Workspace.objects.filter(type=Workspace.Types.ROOT).first()` (e.g., `test_update_workspace_success`, `test_move_workspace_success`) will get `None` instead of a workspace object, likely causing `AttributeError` or test failures.

**Original (correct) order:**
```python
V2TenantBootstrapService(NoopReplicator()).bootstrap_tenant(self.tenant)  # runs full bootstrap, creates TenantMapping
TenantMapping.objects.update_or_create(tenant=self.tenant, defaults={"v2_write_activated_at": timezone.now()})  # sets activation timestamp
```

**New (broken) order:**
```python
TenantMapping.objects.create(tenant=self.tenant, v2_write_activated_at=timezone.now())  # creates TenantMapping first
V2TenantBootstrapService(NoopReplicator()).bootstrap_tenant(self.tenant)  # finds existing TenantMapping, skips bootstrap
```

**Fix**: Restore the original order: run `bootstrap_tenant()` first, then update the `TenantMapping` with `v2_write_activated_at`. The removal of the `_save_replication_event` mock is fine since `NoopReplicator` doesn't need it.

---

**File:** `rbac/management/mcp_views.py` (line 1)

**Low**: Issue 1: Unused `user_is_org_admin` variable (Minor / Bug Risk)

**File:** `rbac/management/mcp_views.py`
**Lines:** ~3618-3622 (in the recommendation section)

The `user_is_org_admin` variable is computed and used in the `user_status` string builder, but the test `test_guide_user_access_delegation_user_not_org_admin` checks for `is_org_admin` in the recommendation text. The actual behavior is correct: when the BOP proxy returns `is_org_admin: False`, the recommendation will contain the string "is_org_admin flag is false". However, when the proxy call fails or org_id is unavailable, `user_is_org_admin` becomes `None` and the `user_status` falls through to the "error" branch, which does NOT include the string "is_org_admin". This means the test `test_guide_user_access_delegation_user_not_org_admin` may be fragile -- it depends on `BYPASS_BOP_VERIFICATION=True` producing a user_info dict that doesn't contain an error key but does have `is_org_admin`.

**Severity:** Low -- the logic is sound, but the test assertion is fragile.

---



---

**File:** `rbac/management/mcp_views.py` (line 1)

**Low**: Issue 2: No input sanitization on `username` parameter (Minor / Consistency)

**File:** `rbac/management/mcp_views.py`
**Lines:** ~3452

The `username` parameter is passed directly into ORM queries (`username__iexact=username`) and into `PrincipalProxy.request_filtered_principals([username])`. While Django ORM parameterizes queries (preventing SQL injection), the username is also interpolated into user-facing strings in the recommendation text via f-strings. An empty string or extremely long username would produce confusing output. Other tools in this file follow the same pattern (no stripping), so this is consistent but worth noting.

**Severity:** Low -- consistent with existing patterns.

---



---

**File:** `tests/management/test_mcp_views.py` (line 1)

**Medium**: Issue 3: The MCPUpdateToolsV2Tests setUp changes remove the `_save_replication_event` mock (Medium / Potential Test Breakage)

**File:** `tests/management/test_mcp_views.py`
**Lines:** ~5295-5320 (diff hunk at end)

The PR modifies `MCPUpdateToolsV2Tests.setUp()` to:
1. Remove the `_save_replication_event` mock entirely
2. Change `TenantMapping.objects.update_or_create` to `TenantMapping.objects.create`
3. Reorder statements so Kessel mocks come before TenantMapping/bootstrap

The `_save_replication_event` mock is used in other V2 test classes to prevent actual outbox writes during tenant bootstrap. Removing it may cause `V2TenantBootstrapService.bootstrap_tenant()` to attempt real outbox writes and fail if the outbox infrastructure isn't available in the test environment. This change was introduced in commit `377f49be3` ("fix(test): remove duplicate TenantMapping creation in V2 tests").

However, since the MCPUpdateToolsV2Tests class doesn't exist on `master` (it comes from PR #2877 which is included in this branch), the changes to this class are not part of the diff against master. The GitHub diff shows them because the PR branch merged #2877 and then modified it. This is an artifact of the branch history, not a problem specific to this PR's feature.

**Severity:** Medium -- potential test breakage, but may be intentional if tests pass.

---



---

**File:** `rbac/management/mcp_views.py` (line 1)

**Minor**: Issue 5: Access permissions query is not tenant-scoped

**File:** `rbac/management/mcp_views.py`
**Line:** ~3543

```python
permissions_list = list(Access.objects.filter(role=role).values_list("permission__permission", flat=True))
```

The Access query filters only by `role=role` without a tenant filter. Since the role is already scoped to the `public` tenant, and Access objects for system roles are also on the public tenant, this is correct. However, it differs from the project convention of always filtering by tenant. Since the role is from the public tenant, adding `.filter(tenant=public_tenant)` would be more explicit and consistent.

**Severity:** Very Low -- functionally correct but could be more explicit.

---

**Comment by lpichler** [comment]
CI Failure:
```
GuideUserAccessDelegationTests.test_guide_user_access_delegation_nonexistent_user)
Edge case: guide_user_access_delegation handles non-existent user gracefully. ... FAIL
 
======================================================================
FAIL: test_guide_user_access_delegation_nonexistent_user (tests.management.test_mcp_views.MCPGuideUserAccessDelegationTests.test_guide_user_access_delegation_nonexistent_user)
Edge case: guide_user_access_delegation handles non-existent user gracefully.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib64/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib64/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/var/workdir/tests/management/test_mcp_views.py", line 5580, in test_guide_user_access_delegation_nonexistent_user
    self.assertIn("error", tool_output["user_info"])
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/unittest/case.py", line 1152, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/unittest/case.py", line 715, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: 'error' not found in {'username': 'nonexistent_user', 'is_org_admin': False, 'is_active': True}
 
----------------------------------------------------------------------
```

**Review by lpichler — APPROVED** [review]
